openshift-knative-operator/pkg/monitoring/monitoring.go

```diff
@@ -21,45 +22,73 @@ import (
 )
 
 const (
-	EnableMonitoringEnvVar       = "ENABLE_SERVING_MONITORING_BY_DEFAULT"
-	EnableMonitoringLabel        = "openshift.io/cluster-monitoring"
-	ObservabilityCMName          = "observability"
-	ObservabilityBackendKey      = "metrics.backend-destination"
-	OpenshiftMonitoringNamespace = "openshift-monitoring"
-	prometheusRoleName           = "knative-serving-prometheus-k8s"
-	prometheusClusterRoleName    = "rbac-proxy-metrics-prom"
-	servingSMRbacManifestPath    = "SERVING_SERVICE_MONITOR_RBAC_MANIFEST_PATH"
+	EnableServingMonitoringEnvVar  = "ENABLE_SERVING_MONITORING_BY_DEFAULT"
+	EnableEventingMonitoringEnvVar = "ENABLE_EVENTING_MONITORING_BY_DEFAULT"
+	EnableMonitoringLabel          = "openshift.io/cluster-monitoring"
+	ObservabilityCMName            = "observability"
+	ObservabilityBackendKey        = "metrics.backend-destination"
+	OpenshiftMonitoringNamespace   = "openshift-monitoring"
+	prometheusRoleName             = "knative-prometheus-k8s"
+	prometheusClusterRoleName      = "rbac-proxy-metrics-prom"
+	smRbacManifestPath             = "SERVICE_MONITOR_RBAC_MANIFEST_PATH"
+)
+
+type KComponentType string
+
+const (
+	Serving  KComponentType = "Serving"
+	Eventing KComponentType = "Eventing"
 )
 
 var (
-	servingComponents = sets.NewString("activator", "autoscaler", "autoscaler-hpa", "controller", "domain-mapping", "domainmapping-webhook", "webhook")
+	servingComponents  = sets.NewString("activator", "autoscaler", "autoscaler-hpa", "controller", "domain-mapping", "domainmapping-webhook", "webhook")
+	eventingComponents = sets.NewString("eventing-controller", "eventing-webhook", "imc-controller", "imc-dispatcher", "mt-broker-controller", "mt-broker-filter", "mt-broker-ingress", "sugar-controller")
 )
 
 func init() {
 	_ = monitoringv1.AddToScheme(scheme.Scheme)
 }
 
-func ReconcileServingMonitoring(ctx context.Context, api kubernetes.Interface, ks *v1alpha1.KnativeServing) error {
-	backend := ks.Spec.CommonSpec.Config[ObservabilityCMName][ObservabilityBackendKey]
-	if shouldEnableMonitoring(backend) {
-		if err := reconcileMonitoringLabelOnNamespace(ctx, ks.Namespace, api, true); err != nil {
+func ReconcileMonitoringForServing(ctx context.Context, api kubernetes.Interface, config v1alpha1.ConfigMapData, spec *v1alpha1.CommonSpec, ns string) error {
```

Comment:
Why pass both ConfigMapData and the CommonSpec? Can't the former be taken off the latter easily?

Reply:
I suppose, this is a confusing struct hierarchy.